src/System.Device.Gpio/System/Device/Gpio/Drivers/LibGpiodDriver.Linux.cs

```diff
@@ -35,20 +50,94 @@ public LibGpiodDriver()
+        private SafeLineHandle SubscribeForEvent(int pinNumber, PinEventTypes eventType)
+        {
+            int eventSuccess = -1; 
+            SafeLineHandle pinHandle = Interop.GetChipLineByOffset(_chip, pinNumber);
+            if (pinHandle != null)
+            {
+                if (eventType.HasFlag(PinEventTypes.Rising) || eventType.HasFlag(PinEventTypes.Falling))
+                {
+                    eventSuccess = Interop.RequestBothEdgesEventForLine(pinHandle, $"Listen {pinNumber} for both edge event");
+                }
+                else
+                {
+                    throw ExceptionHelper.GetArgumentException(ExceptionResource.InvalidEventType);
+                }
+            }
+
+            if (eventSuccess < 0)
+            {
+                throw ExceptionHelper.GetIOException(ExceptionResource.RequestEventError, pinNumber, Marshal.GetLastWin32Error());
+            }
+
+            return pinHandle;
+        }
+
+        private Task InitializeEventDetectionTask(LibGpiodDriverEventHandler eventHandler, CancellationToken token)
+        {
+            return Task.Run(() =>
+            {       
+                while (!(token.IsCancellationRequested || _disposing))
+                {   
+                    // WaitEventResult can be TimedOut, EventOccured or Error, in case of TimedOut will continue waiting
+                    WaitEventResult waitResult = Interop.WaitForEventOnLine(eventHandler.PinHandle);
+                    if (waitResult == WaitEventResult.Error)
+                    {
+                        throw ExceptionHelper.GetIOException(ExceptionResource.EventWaitError, Marshal.GetLastWin32Error(), eventHandler.PinNumber);
+                    }
+
+                    if (waitResult == WaitEventResult.EventOccured)
+                    {
+                        int readResult = Interop.ReadEventForLine(eventHandler.PinHandle);
+                        if (readResult == -1)
+                        {
+                            throw ExceptionHelper.GetIOException(ExceptionResource.EventReadError, Marshal.GetLastWin32Error());
+                        }
+
+                        PinEventTypes eventType = (readResult == 1) ? PinEventTypes.Rising : PinEventTypes.Falling;
+                        eventHandler?.OnPinValueChanged(new PinValueChangedEventArgs(eventType, eventHandler.PinNumber), eventType);
+                    }
+                }
+            }, token);
+        }
+
         protected internal override void AddCallbackForPinValueChangedEvent(int pinNumber, PinEventTypes eventType, PinChangeEventHandler callback)
         {
-            throw new NotImplementedException();
+            if (!_pinNumberToEventHandler.TryGetValue(pinNumber, out LibGpiodDriverEventHandler eventHandler))
+            {
+                eventHandler = new LibGpiodDriverEventHandler(pinNumber, new CancellationTokenSource());
+                if (_pinNumberToEventHandler.TryAdd(pinNumber, eventHandler))
```

Comment:
I think you should put the next two lines in the constructor of the event handler and in here you can use lambda to generate value only when it is not present:
https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.getoradd?view=netframework-4.7.2#System_Collections_Concurrent_ConcurrentDictionary_2_GetOrAdd__0_System_Func__0__1__